Issue 2041

Nastsi opened the pull request:
Deserialization JSON from body of Post-request

Nastsi committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into issue-2041

Nastsi committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into issue-2041

Nastsi committed:
issue#2041. POST

1. Manipulations with "TargetUrl" in JSON
2. Deserialization JSON from POST
3. Fixing jsons in examples

102 commented on RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:
```diff
@@ -37,10 +40,13 @@ public string Get(int id)
             return "value";
         }
-
+        
```
remove trailing spaces

mvahrameev commented on RedirectApplication/RedirectApplication/Json/Json.cs:
```diff
@@ -16,12 +16,20 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
+            {
+                return reader.Value;
+            }
+            else
```
`else` is redundant

Nastsi committed:
issue#2041. Remark correction

Nastsi committed:
issue#2041. Deleting empty line

maxmalov commented on RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:
```diff
@@ -39,8 +42,11 @@ public string Get(int id)
         {
+            var someText = request.Content.ReadAsStringAsync().Result;
+            var reader = new JsonTextReader(new StringReader(someText));
+            PostJson content = JsonSerializer.CreateDefault().Deserialize<PostJson>(reader);
```
This method isn't doing anything actually

Nastsi replied:
In this method I get JSON from Post-request and deserialize it in PostJson object

maxmalov replied:
aaaaand then...?

Nastsi replied:
What happened then will be written in another tickets. This one is about getting JSON and deserialization

maxmalov replied:
So again, it should be in another tickets PRs https://github.com/hwdtech/redirect/pull/7#discussion_r84063406

maxmalov commented on RedirectApplication/RedirectApplication/Json/Json.cs:
```diff
@@ -16,8 +16,13 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
 
         public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
         {
+            if (reader.Path == "TargetUrl")
```
Looks like a hack. Can you elaborate a little?

Nastsi replied:
Example of JSON:

{
"TargetUrl":"localhost",
"Conditions": [ {"Name:":"ByBrowser","Browser":"Chrome","Url":"http://pikabu.ru/"} ]
}

Either we have just one field "TargetUrl", either we working with everything in "Conditions" (that may look like a tree).
I just don't know how to complicate this

102 replied:
@Nastsi Use markdown syntax to insert code examples into comments, i.e.:

``` json
{
  "TargetUrl":"localhost",
  "Conditions": [ {"Name:":"ByBrowser","Browser":"Chrome","Url":"http://pikabu.ru/"} ]
}
```

will be generated from

``````
```json
{
  "TargetUrl":"localhost",
  "Conditions": [ {"Name:":"ByBrowser","Browser":"Chrome","Url":"http://pikabu.ru/"} ]
}
``` 
``````

maxmalov commented:
As was sad earlier: https://github.com/hwdtech/redirect/pull/7#issuecomment-255125749

Nastsi committed:
issue#2041 Separate code in Post-method and ReadJson-method to different methods

For "Post" have been created method "Deseriliazion" with getting Json
from body of request and deserialization it in PostJson.
For "ReadJson" have been created TargetDeserialization and
ConditionDeserialization

Nastsi committed:
issue#2041. Deleting excess JArray.Load

maxmalov commented on RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:
```diff
@@ -39,8 +42,16 @@ public string Get(int id)
+            Deserialization(request);
+        }
+
+        public void Deserialization(HttpRequestMessage request)
```
I was talking about such schema:

``` csharp
public void Post(...)
{
  // var content = DeserializeRecord(...);
}

private PostJson DeserializeRecord(HttpRequestMessage request)
{
  // return deserialized record
}
```

Nastsi replied:
I did this in the first place but then rethink

Nastsi committed:
issue#2041 Make method "Deserialization" return a value

jacob-l reviewed (approved):
We can merge it now. Later I will ask to change several things
